components/ws-manager-api/typescript/src/client-provider.ts

```diff
@@ -39,41 +32,48 @@ export class WorkspaceManagerClientProvider implements Disposable {
+     * If starting a workspace fails in one cluster, the caller is expected to "pop" another cluster
+     * of the list until the workspace start has succeeded or pop returns undefined.
      *
-     * @returns The WorkspaceManagerClient that was chosen to start the next workspace with.
+     * @param user user who wants to starts a workspace manager
+     * @param workspace the workspace we want to start
+     * @param instance the instance we want to start
+     * @returns a set of workspace clusters we can start the workspace in
      */
-    public async getStartManager(user: ExtendedUser, workspace: Workspace, instance: WorkspaceInstance, exceptInstallations?: string[]): Promise<{ manager: PromisifiedWorkspaceManagerClient, installation: string }> {
-        let availableCluster = await this.getAvailableStartCluster(user, workspace, instance);
-        if (!!exceptInstallations) {
-            availableCluster = availableCluster.filter(c => !exceptInstallations?.includes(c.name));
-        }
-
-        let chosenCluster = chooseCluster(availableCluster.filter(admissionPreferencesFilter(user)))
-        if (!chosenCluster) {
-            // We did not choose a prefered cluster. That's ok, we'll try with the remaining set.
-            chosenCluster = chooseCluster(availableCluster);
-        }
-        if (!chosenCluster) {
-            // We still haven't choosen a cluster. That's a problem.
-            throw new Error("no available workspace cluster to choose from!");
-        }
+    public async getStartClusterSets(user: ExtendedUser, workspace: Workspace, instance: WorkspaceInstance): Promise<IWorkspaceClusterStartSet> {
+        const allClusters = await this.source.getAllWorkspaceClusters();
+        const availableClusters = allClusters.filter(c => c.score >= 0 && c.state === "available");
 
+        const sets = workspaceClusterSets.map(constraints => {
+            const r = constraints.constraint(availableClusters, user, workspace, instance);
+            if (!r) {
+                return;
+            }
+            return new ClusterSet(this, r);
+        }).filter(s => s !== undefined) as ClusterSet[];
 
-        const grpcOptions: grpc.ClientOptions = {
-            ...defaultGRPCOptions,
-        };
-        const client = await this.get(chosenCluster.name, grpcOptions);
         return {
-            manager: client,
-            installation: chosenCluster.name,
-        };
-    }
-
-    public async getAvailableStartCluster(user: User, workspace: Workspace, instance: WorkspaceInstance): Promise<WorkspaceClusterWoTLS[]> {
-        const allClusters = await this.source.getAllWorkspaceClusters();
-        const availableClusters = allClusters.filter(c => c.score >= 0 && c.state === "available").filter(admissionConstraintsFilter(user, workspace, instance));
-        return availableClusters;
+            [Symbol.asyncIterator]: (): AsyncIterator<ClusterClientEntry> => {
```

Comment:
I find the asyncIterator syntax and usage rather noisy and hard to read/follow, mostly because it's so uncommon.
I assume it's used here to hide the lazy initialization of clients/filtering of used clients here. Not sure I would have gone the same path, but fine with me as is.

Reply:
Same here - it's a rather unsightly construct. Originally I used a custom "pop" function, but thought that moving to something more common is the way to go. Not sure that was worth the trouble though.

The purpose is indeed to facilitate lazy init of the clients.